subprojects/packagefiles/openssl/meson.build

```diff
@@ -0,0 +1,457 @@
+    arch_subdir = 'linux-ppc64le'
+  elif host_machine.cpu_family() == 'ppc64' and host_machine.system() == 'linux'
+    arch_subdir = 'linux-ppc64'
+  elif host_machine.cpu_family() == 's390x' and host_machine.system() == 'linux'
+    arch_subdir = 'linux64-s390x'
+  elif host_machine.cpu_family() == 'arm' and host_machine.system() in ['linux', 'android']
+    arch_subdir = 'linux-armv4'
+  elif host_machine.cpu_family() == 'aarch64' and host_machine.system() in ['linux', 'android']
+    arch_subdir = 'linux-aarch64'
+  elif host_machine.cpu_family() == 'x86' and host_machine.system() in ['dragonfly', 'freebsd', 'netbsd', 'openbsd']
+    arch_subdir = 'BSD-x86'
+  elif host_machine.cpu_family() == 'x86' and host_machine.system() in ['linux', 'android']
+    arch_subdir = 'linux-elf'
+  elif host_machine.cpu_family() == 'x86' and host_machine.system() == 'darwin'
+    arch_subdir = 'darwin-i386-cc'
+  elif host_machine.cpu_family() == 'x86' and host_machine.system() == 'sunos'
+    arch_subdir = 'solaris-x86-gcc'
+  elif host_machine.cpu_family() == 'x86' and host_machine.system() == 'windows'
+    arch_subdir = 'VC-WIN32'
+  elif host_machine.cpu_family() == 'x86_64' and host_machine.system() in ['dragonfly', 'freebsd', 'netbsd', 'openbsd']
+    arch_subdir = 'BSD-x86_64'
+  elif host_machine.cpu_family() == 'x86_64' and host_machine.system() == 'darwin'
+    arch_subdir = 'darwin64-x86_64-cc'
+  elif host_machine.cpu_family() == 'aarch64' and host_machine.system() == 'darwin'
+    arch_subdir = 'darwin64-arm64-cc'
+  elif host_machine.cpu_family() == 'x86_64' and host_machine.system() == 'sunos'
+    arch_subdir = 'solaris64-x86_64-gcc'
+  elif host_machine.cpu_family() == 'x86_64' and host_machine.system() == 'windows'
+    arch_subdir = 'VC-WIN64A'
+  elif host_machine.cpu_family() == 'aarch64' and host_machine.system() == 'windows'
+    arch_subdir = 'VC-WIN64-ARM'
+  elif host_machine.cpu_family() == 'x86_64' and host_machine.system() == 'linux'
+    arch_subdir = 'linux-x86_64'
+  elif host_machine.cpu_family() == 'mips64' and host_machine.system() == 'linux'
+    arch_subdir = 'linux64-mips64'
+  else
+    error('Unsupported arch+OS combo: ' + host_machine.cpu_family() + ' + ' + host_machine.system())
+  endif
+
+  message('OpenSSL is configured without ASM support')
+else
+  message('OpenSSL is configured with ASM support')
+endif
+
+subdir('generated-config/archs' / arch_subdir / asm)
+
+# Build options specific to OS
+if host_machine.system() == 'windows'
+  defines += [
+    ## default of Win. See INSTALL in openssl repo.
+    'OPENSSLDIR="C:\\Program Files\\Common Files\\SSL"',
+    'ENGINESDIR="NUL"',
+    'OPENSSL_SYS_WIN32', 'WIN32_LEAN_AND_MEAN', 'L_ENDIAN',
+    '_CRT_SECURE_NO_DEPRECATE', 'UNICODE', '_UNICODE',
+  ]
+  if compiler.get_id() == 'msvc'
+    c_args += [
+      '-wd4090', '-Gs0', '-GF', '-Gy', '-nologo',
+    ]
+    foreach library: openssl_libraries
+      dependencies += [
+        compiler.find_library(library),
```

Comment:
Why only if the compiler is msvc, do you perform find_library without `static: get_option('default_library') == 'static'`?

Also there is a difference between wanting to have a static libssl.a, and wanting to statically link all your dependencies too. In my project(s) I use a meson_options.txt option "buildstatic", fluidsynth uses "try-static-deps" for this.

Reply:
It fails under MSVC, here is an example CI run I just did to reproduce it: https://github.com/nazar-pc/mediasoup/runs/3306203420?check_suite_focus=true:

```
Executing subproject openssl 

openssl| Project name: openssl
openssl| Project version: 1.1.1k
openssl| C compiler for the host machine: cl (msvc 19.29.30040 "Microsoft (R) C/C++ Optimizing Compiler Version 19.29.30040 for x64")
openssl| C linker for the host machine: link link 14.29.30040.0
openssl| Library ws2_32 found: YES
openssl| Library gdi32 found: YES
openssl| Library advapi32 found: YES
openssl| Library crypt32 found: YES
openssl| Library user32 found: YES

subprojects\openssl\meson.build:45:2: ERROR: C static library 'ws2_32' not found

A full log can be found at D:\a\mediasoup\mediasoup\worker\out\Release\meson-logs\meson-log.txt
mingw32-make: *** [Makefile:47: setup] Error 1
mingw32-make: Leaving directory 'D:/a/mediasoup/mediasoup/worker'
```

:man_shrugging: 

I care so little about Windows that I just decided to leave it as it is right now and leave an opportunity to improve it for someone else.

Reply:
I... don't understand the error? Where does it even come from?

meson treats the static kwarg as "prefer static", not "require static". So it should just find a shared one instead.

Reply:
Line 45 is here: https://github.com/nazar-pc/mediasoup/blob/f28cffd4b6dae60cd97615554ac1d3a5048ad9e9/worker/subprojects/openssl/meson.build#L45-L66
```
  dependencies += [
    compiler.find_library(
      'ws2_32',
      static: get_option('default_library') == 'static',
    ),
    compiler.find_library(
      'gdi32',
      static: get_option('default_library') == 'static',
    ),
    compiler.find_library(
      'advapi32',
      static: get_option('default_library') == 'static',
    ),
    compiler.find_library(
      'crypt32',
      static: get_option('default_library') == 'static',
    ),
    compiler.find_library(
      'user32',
      static: get_option('default_library') == 'static',
    ),
  ]
```

Reply:
Could we at least do this lookup once then, not repeated N times for every system and compiler?

We can just do

```
try_static_deps = get_option('try-static-deps') and  compiler.get_id() != 'msvc'
compiler.find_library(library, static: try_static_deps)
```

(or condense this to one line)

Reply:
Also I was not asking "where is line 45". :p

Reply:
I'm not sure it is worth having an option just for this, but otherwise refactored to have just one library lookup, hopefully it is fine now

Reply:
Good enough I suppose...

Ideally this option would be a core meson option.